src/EventStore.Core/Services/Transport/Grpc/Enumerators.AllSubscription.cs

```diff
@@ -52,465 +54,314 @@ public class AllSubscription : ISubscriptionEnumerator {
 
-					if (_buffer.TryDequeue(out current)) {
-						_current = current;
-						_currentPosition = Position.FromInt64(current.OriginalPosition.Value.CommitPosition,
-							current.OriginalPosition.Value.PreparePosition);
-						return true;
-					}
+								await _channel.Writer.WriteAsync(@event, ct).ConfigureAwait(false);
+							}
 
-					if (isEnd) {
-						return false;
-					}
+							if (completed.IsEndOfStream) {
+								GoLive(Position.FromInt64(
+									completed.NextPos.CommitPosition,
+									completed.NextPos.PreparePosition));
+								return;
+							}
 
-					if (_disposedTokenSource.IsCancellationRequested) {
-						return false;
+							ReadPage(Position.FromInt64(
+								completed.NextPos.CommitPosition,
+								completed.NextPos.PreparePosition), OnMessage);
+							return;
+						case ReadAllResult.AccessDenied:
+							_channel.Writer.TryComplete(RpcExceptions.AccessDenied());
+							return;
+						default:
+							_channel.Writer.TryComplete(RpcExceptions.UnknownError(completed.Result));
+							return;
 					}
+				}
+			}
 
-					goto ReadLoop;
+			private void GoLive(Position startPosition) {
+				var liveEvents = Channel.CreateBounded<ResolvedEvent>(BoundedChannelOptions);
+				var caughtUpSource = new TaskCompletionSource<TFPos>();
+				var liveMessagesCancelled = 0;
 
-					void OnMessage(Message message) {
-						if (message is ClientMessage.NotHandled notHandled &&
-						    RpcExceptions.TryHandleNotHandled(notHandled, out var ex)) {
-							readNextSource.TrySetException(ex);
-							return;
-						}
+				Log.Information(
+					"Live subscription {subscriptionId} to $all running from {position}...",
+					_subscriptionId, startPosition);
 
-						if (!(message is ClientMessage.ReadAllEventsForwardCompleted completed)) {
-							readNextSource.TrySetException(
-								RpcExceptions.UnknownMessage<ClientMessage.ReadAllEventsForwardCompleted>(message));
-							return;
-						}
+				_bus.Publish(new ClientMessage.SubscribeToStream(Guid.NewGuid(), _subscriptionId,
+					new ContinuationEnvelope(OnSubscriptionMessage, _semaphore, _cancellationToken), _subscriptionId,
+					string.Empty, _resolveLinks, _user));
 
-						switch (completed.Result) {
-							case ReadAllResult.Success:
-								foreach (var @event in completed.Events) {
-									var position = Position.FromInt64(
-										@event.OriginalPosition.Value.CommitPosition,
-										@event.OriginalPosition.Value.PreparePosition);
-									if (position <= _startPosition) {
-										Log.Verbose(
-											"Catch-up subscription {subscriptionId} to $all skipped event {position}.",
-											_subscriptionId, position);
-										continue;
-									}
-
-									Log.Verbose(
-										"Catch-up subscription {subscriptionId} to $all received event {position}.",
-										_subscriptionId, position);
-
-									_buffer.Enqueue(@event);
-								}
+				Task.Factory.StartNew(PumpLiveMessages, _cancellationToken);
```

Comment:
Should we start this task only after the subscription is confirmed? (e.g access could be denied)